### DISCUSSION TITLE
HIGH: Allow choosing email while queueing up PR to merge

### DISCUSSION THREAD

**Discussion by purajit** [community/community#64785] [Pull Requests]
IMO this is quite critical.

# Situation

Currently, we have some cases where the commit author email is being overwritten with the account email during merge - meaning, when people set their `.gitconfig` locally to a specific email, the PR commit contains the correct email, but then the final commit pushed to the branch is associated with a different one when using a merge queue.

After some testing, I found out that this is because the user likely in the past made a GitHub UI action on that repo using that other email, and now `gh pr merge` or clicking the Big Green Button uses that email for all operations.

This also means that https://docs.github.com/en/github-ae@latest/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address is an incorrect document, since setting your gitconfig is _not_ guaranteed to be enough, and GitHub can decide not to honor this setting.

Another odd concern here is that the author email gets overwritten with `viewerDefaultCommitEmail` _even for CLI-triggered merges_, which IMO is unexpected, since it's meant for UI operations only.

# Problem

There seems to be no way to change the email used on that repo - without a merge queue, you get the option of choosing which email to use. With merge queue, there is no such drop down. I understand that I can probably ask the user to create a PR from the UI and use that do reassociate the repo with the correct email, but that seems like a very awkward workaround, and a very awkward thing to ask a user to do on every repo they have this problem with.

# Possible solutions

* Author emails explicitly provided should never be overwritten
* Add the same email dropdown for the "Merge when ready" button
* Allow viewing/modifying per-repo emails all at once in settings. Or even via API.
* Org-specific email settings (I know this one has already been asked for)

I understand the latter suggestions are much bigger considerations, but I at least hope that adding the dropdown like in https://github.blog/changelog/2022-09-15-git-commit-author-shown-when-squash-merging-a-pull-request would be reasonable to accomplish rather soon.

Without this, we are unable to enforce repo rules that restrict commit author emails.

**Comment by Jaber-Saed** [comment]
Based on the provided text, it seems you're discussing an issue related to Git commit author emails and their handling during the merge process on GitHub. The situation involves the overwrite of commit author emails during a merge, which can lead to inconsistencies between the PR commit and the final commit pushed to the branch. This occurs when users have set their local .gitconfig to a specific email address, and actions performed through the GitHub UI use a different email address.

The problem lies in the inability to change the email associated with a repository, particularly when using a merge queue. This limitation can result in difficulties enforcing repo rules that restrict commit author emails.

Here are the possible solutions you've identified:

Add Email Dropdown for "Merge when ready" Button: Introduce an email dropdown option similar to what's available in the context of squashing commits during a merge. This would allow users to choose the desired email when using the "Merge when ready" button.

Per-Repo Email Management: Enable the ability to view and modify per-repo email settings in a consolidated manner, either through the settings UI or an API. This would provide users with more control over the email associated with specific repositories.

Org-Specific Email Settings: Implement organization-wide email settings, allowing administrators to define email preferences for repositories within the organization. This feature could align with broader org-specific settings that have been previously requested.

While the latter suggestions may involve more complex development efforts, implementing the first solution—adding an email dropdown for the "Merge when ready" button—could potentially provide a more immediate resolution to the issue.

Overall, the problem of commit author email inconsistencies during merges is affecting repository integrity and rule enforcement. The proposed solutions aim to address this issue and provide users with greater control over their commit author information. Considering the importance of maintaining accurate commit authorship records, implementing one or more of these solutions would likely be well-received by the GitHub community.

**Reply by purajit** [reply]
Reporting for useless ChatGPT-generated shpiel.

**Comment by cobexer** [comment]
A solution similar to the per-organization notification routing would be preferred!

**Comment by njgrisafi** [comment]
Any update on this? We have this issue in our org right now.

The way we are fixing it is by merging a PR into a non-default branch. That gives you the option to select your commit email, once you merge with the correct email it is fixed in the merge queue.

<img width="896" height="443" alt="2026-05-14_13-56-15" src="https://github.com/user-attachments/assets/35898cc6-1105-4f64-a80f-15807c525a8a" />

**Comment by ejhayes** [comment]
2026 and this problem still exists.
